File: src/services/projectService.js
const fs = require('fs-extra');
const path = require('path');
const { CLAUDE_PATHS } = require('../config/paths');
const { AppError } = require('../middleware/errorHandler');
const { calculateUsageMetrics } = require('./pricingService');
const { processProjectDataWithRust, isRustProcessorAvailable } = require('./rustProcessor');

/**
 * Node.js実装のプロジェクトデータ処理
 */
async function processProjectDataNodeJS() {
  try {
    if (!(await fs.pathExists(CLAUDE_PATHS.projects))) {
      console.log('Projects directory not found');
      return { dailyUsage: [], monthlyUsage: [], modelUsage: [], projects: [] };
    }

    const projectDirs = await fs.readdir(CLAUDE_PATHS.projects);
    console.log(`Found ${projectDirs.length} project directories`);
    
    const usageByDate = new Map();
    const usageByMonth = new Map();
    const usageByModel = new Map();
    const projects = [];
    const detailedUsage = []; // Store detailed entries with timestamps
    
    // 並列でプロジェクトディレクトリを処理
    await Promise.all(projectDirs.map(async (projectDir) => {
      const projectPath = path.join(CLAUDE_PATHS.projects, projectDir);
      const stats = await fs.stat(projectPath);
      
      if (!stats.isDirectory()) return;
      
      const files = await fs.readdir(projectPath);
      let totalTokens = 0;
      let totalCost = 0;
      let messageCount = 0;
      let lastActivity = null;
      
      // 並列でJSONLファイルを処理
      await Promise.all(files.map(async (file) => {
        if (!file.endsWith('.jsonl')) return;
        
        const filePath = path.join(projectPath, file);
        const content = await fs.readFile(filePath, 'utf8');
        const lines = content.split('\n').filter(line => line.trim());
        
        for (const line of lines) {
          try {
            const data = JSON.parse(line);
            messageCount++;
            
            if (data.timestamp) {
              const timestamp = new Date(data.timestamp);
              if (!lastActivity || timestamp > lastActivity) {
                lastActivity = timestamp;
              }
            }
            
            if (data.message && data.message.usage) {
              const usage = data.message.usage;
              const timestamp = data.timestamp;
              const model = data.message.model;

              const metrics = calculateUsageMetrics(usage, model);

              // Store detailed entry with timestamp for time-based filtering
              if (timestamp) {
                detailedUsage.push({
                  timestamp,
                  sessionId: data.sessionId,
                  model,
                  inputTokens: metrics.inputTokens,
                  outputTokens: metrics.outputTokens,
                  cachedTokens: metrics.cachedTokens,
                  totalTokens: metrics.totalTokens,
                  cost: metrics.cost,
                  newInputTokens: metrics.newInputTokens || 0,
                  cacheCreationTokens: metrics.cacheCreationTokens || 0,
                  cacheReadTokens: metrics.cacheReadTokens || 0
                });
              }

              // 日毎データ
              if (timestamp) {
                const date = new Date(timestamp).toISOString().split('T')[0];
                if (!usageByDate.has(date)) {
                  usageByDate.set(date, {
                    date,
                    inputTokens: 0,
                    outputTokens: 0,
                    cachedTokens: 0,
                    totalTokens: 0,
                    cost: 0,
                    sessions: new Set(),
                    // Detailed breakdown
                    newInputTokens: 0,
                    cacheCreationTokens: 0,
                    cacheReadTokens: 0
                  });
                }
                
                const dayData = usageByDate.get(date);
                dayData.inputTokens += metrics.inputTokens;
                dayData.outputTokens += metrics.outputTokens;
                dayData.cachedTokens += metrics.cachedTokens;
                dayData.totalTokens += metrics.totalTokens;
                dayData.cost += metrics.cost;
                
                // Add detailed breakdown
                dayData.newInputTokens += metrics.newInputTokens || 0;
                dayData.cacheCreationTokens += metrics.cacheCreationTokens || 0;
                dayData.cacheReadTokens += metrics.cacheReadTokens || 0;
                
                if (data.sessionId) {
                  dayData.sessions.add(data.sessionId);
                }
                
                // 月毎データ
                const monthKey = `${new Date(timestamp).getFullYear()}-${String(new Date(timestamp).getMonth() + 1).padStart(2, '0')}`;
                if (!usageByMonth.has(monthKey)) {
                  usageByMonth.set(monthKey, {
                    month: monthKey,
                    inputTokens: 0,
                    outputTokens: 0,
                    cachedTokens: 0,
                    totalTokens: 0,
                    cost: 0,
                    sessions: new Set(),
                    messages: 0,
                    // Detailed breakdown
                    newInputTokens: 0,
                    cacheCreationTokens: 0,
                    cacheReadTokens: 0
                  });
                }
                
                const monthData = usageByMonth.get(monthKey);
                monthData.inputTokens += metrics.inputTokens;
                monthData.outputTokens += metrics.outputTokens;
                monthData.cachedTokens += metrics.cachedTokens;
                monthData.totalTokens += metrics.totalTokens;
                monthData.messages++;
                monthData.cost += metrics.cost;
                
                // Add detailed breakdown
                monthData.newInputTokens += metrics.newInputTokens || 0;
                monthData.cacheCreationTokens += metrics.cacheCreationTokens || 0;
                monthData.cacheReadTokens += metrics.cacheReadTokens || 0;
                
                if (data.sessionId) {
                  monthData.sessions.add(data.sessionId);
                }
              }
              
              // モデル別データ
              if (model) {
                if (!usageByModel.has(model)) {
                  usageByModel.set(model, {
                    model,
                    inputTokens: 0,
                    outputTokens: 0,
                    cachedTokens: 0,
                    totalTokens: 0,
                    cost: 0,
                    messages: 0,
                    sessions: new Set(),
                    // Detailed breakdown for analysis
                    newInputTokens: 0,
                    cacheCreationTokens: 0,
                    cacheReadTokens: 0
                  });
                }
                
                const modelData = usageByModel.get(model);
                modelData.inputTokens += metrics.inputTokens;
                modelData.outputTokens += metrics.outputTokens;
                modelData.cachedTokens += metrics.cachedTokens;
                modelData.totalTokens += metrics.totalTokens;
                modelData.messages++;
                modelData.cost += metrics.cost;
                
                // Add detailed breakdown
                modelData.newInputTokens += metrics.newInputTokens || 0;
                modelData.cacheCreationTokens += metrics.cacheCreationTokens || 0;
                modelData.cacheReadTokens += metrics.cacheReadTokens || 0;
                
                if (data.sessionId) {
                  modelData.sessions.add(data.sessionId);
                }
              }
              
              // プロジェクト統計
              totalTokens += metrics.totalTokens;
              totalCost += metrics.cost;
            }
          } catch (e) {
            // JSONパースエラーは無視
          }
        }
      }));
      
      projects.push({
        name: projectDir,
        path: projectPath,
        totalTokens,
        totalCost: totalCost.toFixed(4),
        messageCount,
        lastActivity
      });
    }));
    
    // 結果を配列に変換
    const dailyData = Array.from(usageByDate.values()).map(day => ({
      ...day,
      cost: day.cost.toFixed(4),
      sessions: day.sessions.size
    })).sort((a, b) => new Date(a.date) - new Date(b.date));
    
    const monthlyData = Array.from(usageByMonth.values()).map(month => ({
      ...month,
      cost: month.cost.toFixed(4),
      sessions: month.sessions.size
    })).sort((a, b) => a.month.localeCompare(b.month));
    
    const modelData = Array.from(usageByModel.values()).map(model => ({
      ...model,
      cost: model.cost.toFixed(4),
      sessions: model.sessions.size
    })).sort((a, b) => b.totalTokens - a.totalTokens);
    
    projects.sort((a, b) => new Date(b.lastActivity) - new Date(a.lastActivity));

    // Sort detailed usage by timestamp
    detailedUsage.sort((a, b) => new Date(a.timestamp) - new Date(b.timestamp));

    // 全期間のユニークセッション数を計算
    const allSessions = new Set();
    usageByDate.forEach(day => {
      day.sessions.forEach(sessionId => allSessions.add(sessionId));
    });

    return {
      dailyUsage: dailyData,
      monthlyUsage: monthlyData,
      modelUsage: modelData,
      projects,
      detailedUsage,
      totalSessions: allSessions.size
    };
  } catch (error) {
    console.error('Error processing project data:', error);
    throw new AppError('Failed to process project data', 500, 'PROJECT_PROCESSING_ERROR');
  }
}

/**
 * プロジェクトデータ処理のエントリーポイント
 * Rustプロセッサを優先的に使用し、失敗時はNode.js実装にフォールバック
 */
async function processProjectData() {
  const startTime = Date.now();

  // Rustプロセッサが利用可能かチェック
  if (isRustProcessorAvailable()) {
    console.log('[Hybrid] Attempting to use Rust processor...');
    const rustResult = processProjectDataWithRust();

    if (rustResult) {
      console.log(`[Hybrid] Successfully processed with Rust in ${Date.now() - startTime}ms`);
      return rustResult;
    }

    console.log('[Hybrid] Rust processor failed, falling back to Node.js implementation');
  } else {
    console.log('[Hybrid] Rust processor not available, using Node.js implementation');
  }

  // Node.js実装にフォールバック
  console.log('[Hybrid] Processing with Node.js...');
  const nodeResult = await processProjectDataNodeJS();
  console.log(`[Hybrid] Successfully processed with Node.js in ${Date.now() - startTime}ms`);
  return nodeResult;
}

module.exports = {
  processProjectData,
  processProjectDataNodeJS // テスト用にexport
};

File: rust-backend/src/services/project_service.rs
use crate::models::*;
use anyhow::{Context, Result};
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::PathBuf;

pub fn process_daily_usage(projects_path: &str) -> Result<Vec<DailyUsage>> {
    let project_data = process_all_project_data(projects_path)?;
    Ok(project_data.daily_usage)
}

pub fn process_monthly_usage(projects_path: &str) -> Result<Vec<MonthlyUsage>> {
    let project_data = process_all_project_data(projects_path)?;
    Ok(project_data.monthly_usage)
}

pub fn process_model_usage(projects_path: &str) -> Result<Vec<ModelUsage>> {
    let project_data = process_all_project_data(projects_path)?;
    Ok(project_data.model_usage)
}

pub fn process_projects(projects_path: &str) -> Result<Vec<ProjectData>> {
    let project_data = process_all_project_data(projects_path)?;
    Ok(project_data.projects)
}

pub struct AllProjectData {
    pub daily_usage: Vec<DailyUsage>,
    pub monthly_usage: Vec<MonthlyUsage>,
    pub model_usage: Vec<ModelUsage>,
    pub projects: Vec<ProjectData>,
}

fn process_all_project_data(projects_path: &str) -> Result<AllProjectData> {
    let project_dirs: Vec<PathBuf> = fs::read_dir(projects_path)
        .context("Failed to read projects directory")?
        .filter_map(|entry| entry.ok())
        .map(|entry| entry.path())
        .filter(|path| path.is_dir())
        .collect();

    let mut usage_by_date: HashMap<String, DayData> = HashMap::new();
    let mut usage_by_month: HashMap<String, MonthData> = HashMap::new();
    let mut usage_by_model: HashMap<String, ModelData> = HashMap::new();
    let mut projects: Vec<ProjectInternal> = Vec::new();

    for project_dir in &project_dirs {
        let project_name = project_dir
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("")
            .to_string();

        let files: Vec<PathBuf> = fs::read_dir(project_dir)
            .context(format!("Failed to read project directory: {:?}", project_dir))?
            .filter_map(|entry| entry.ok())
            .map(|entry| entry.path())
            .filter(|path| {
                path.extension()
                    .and_then(|ext| ext.to_str())
                    .map(|ext| ext == "jsonl")
                    .unwrap_or(false)
            })
            .collect();

        let mut project_total_tokens = 0u64;
        let mut project_total_cost = 0.0f64;
        let mut project_message_count = 0usize;
        let mut project_last_activity: Option<String> = None;

        for file in files {
            let content = fs::read_to_string(&file)
                .context(format!("Failed to read file: {:?}", file))?;

            for line in content.lines() {
                let trimmed = line.trim();
                if trimmed.is_empty() {
                    continue;
                }

                let msg: Message = match serde_json::from_str(trimmed) {
                    Ok(m) => m,
                    Err(_) => continue,
                };

                if let (Some(message_content), Some(timestamp)) =
                    (msg.message, msg.timestamp.as_ref())
                {
                    project_message_count += 1;

                    // Update last activity
                    if project_last_activity.is_none()
                        || timestamp > project_last_activity.as_ref().unwrap()
                    {
                        project_last_activity = Some(timestamp.clone());
                    }

                    if let Some(usage) = message_content.usage {
                        let metrics = UsageMetrics::from_usage(&usage);
                        let model = message_content.model.unwrap_or_else(|| "unknown".to_string());

                        project_total_tokens += metrics.total_tokens;
                        project_total_cost += metrics.cost;

                        // Daily data
                        let date = timestamp.split('T').next().unwrap_or("").to_string();
                        let day_data = usage_by_date.entry(date.clone()).or_insert(DayData {
                            date: date.clone(),
                            input_tokens: 0,
                            output_tokens: 0,
                            cached_tokens: 0,
                            total_tokens: 0,
                            cost: 0.0,
                            sessions: HashSet::new(),
                            new_input_tokens: 0,
                            cache_creation_tokens: 0,
                            cache_read_tokens: 0,
                        });

                        day_data.input_tokens += metrics.input_tokens;
                        day_data.output_tokens += metrics.output_tokens;
                        day_data.cached_tokens += metrics.cached_tokens;
                        day_data.total_tokens += metrics.total_tokens;
                        day_data.cost += metrics.cost;
                        day_data.new_input_tokens += metrics.new_input_tokens;
                        day_data.cache_creation_tokens += metrics.cache_creation_tokens;
                        day_data.cache_read_tokens += metrics.cache_read_tokens;

                        if let Some(session_id) = &msg.session_id {
                            day_data.sessions.insert(session_id.clone());
                        }

                        // Monthly data
                        let month = {
                            let parts: Vec<&str> = date.split('-').collect();
                            if parts.len() >= 2 {
                                format!("{}-{}", parts[0], parts[1])
                            } else {
                                continue;
                            }
                        };

                        let month_data =
                            usage_by_month.entry(month.clone()).or_insert(MonthData {
                                month: month.clone(),
                                input_tokens: 0,
                                output_tokens: 0,
                                cached_tokens: 0,
                                total_tokens: 0,
                                cost: 0.0,
                                sessions: HashSet::new(),
                                messages: 0,
                                new_input_tokens: 0,
                                cache_creation_tokens: 0,
                                cache_read_tokens: 0,
                            });

                        month_data.input_tokens += metrics.input_tokens;
                        month_data.output_tokens += metrics.output_tokens;
                        month_data.cached_tokens += metrics.cached_tokens;
                        month_data.total_tokens += metrics.total_tokens;
                        month_data.cost += metrics.cost;
                        month_data.messages += 1;
                        month_data.new_input_tokens += metrics.new_input_tokens;
                        month_data.cache_creation_tokens += metrics.cache_creation_tokens;
                        month_data.cache_read_tokens += metrics.cache_read_tokens;

                        if let Some(session_id) = &msg.session_id {
                            month_data.sessions.insert(session_id.clone());
                        }

                        // Model data
                        let model_data =
                            usage_by_model.entry(model.clone()).or_insert(ModelData {
                                model: model.clone(),
                                input_tokens: 0,
                                output_tokens: 0,
                                cached_tokens: 0,
                                total_tokens: 0,
                                cost: 0.0,
                                sessions: HashSet::new(),
                                messages: 0,
                                new_input_tokens: 0,
                                cache_creation_tokens: 0,
                                cache_read_tokens: 0,
                            });

                        model_data.input_tokens += metrics.input_tokens;
                        model_data.output_tokens += metrics.output_tokens;
                        model_data.cached_tokens += metrics.cached_tokens;
                        model_data.total_tokens += metrics.total_tokens;
                        model_data.cost += metrics.cost;
                        model_data.messages += 1;
                        model_data.new_input_tokens += metrics.new_input_tokens;
                        model_data.cache_creation_tokens += metrics.cache_creation_tokens;
                        model_data.cache_read_tokens += metrics.cache_read_tokens;

                        if let Some(session_id) = &msg.session_id {
                            model_data.sessions.insert(session_id.clone());
                        }
                    }
                }
            }
        }

        projects.push(ProjectInternal {
            name: project_name,
            total_tokens: project_total_tokens,
            total_cost: project_total_cost,
            message_count: project_message_count,
            last_activity: project_last_activity,
        });
    }

    // Convert to response types
    let mut daily_usage: Vec<DailyUsage> = usage_by_date
        .into_iter()
        .map(|(_, day)| DailyUsage {
            date: day.date,
            input_tokens: day.input_tokens,
            output_tokens: day.output_tokens,
            cached_tokens: day.cached_tokens,
            total_tokens: day.total_tokens,
            cost: format!("{:.4}", day.cost),
            sessions: day.sessions.len(),
            new_input_tokens: day.new_input_tokens,
            cache_creation_tokens: day.cache_creation_tokens,
            cache_read_tokens: day.cache_read_tokens,
        })
        .collect();

    daily_usage.sort_by(|a, b| a.date.cmp(&b.date));

    let mut monthly_usage: Vec<MonthlyUsage> = usage_by_month
        .into_iter()
        .map(|(_, month)| MonthlyUsage {
            month: month.month,
            input_tokens: month.input_tokens,
            output_tokens: month.output_tokens,
            cached_tokens: month.cached_tokens,
            total_tokens: month.total_tokens,
            cost: format!("{:.4}", month.cost),
            sessions: month.sessions.len(),
            messages: month.messages,
            new_input_tokens: month.new_input_tokens,
            cache_creation_tokens: month.cache_creation_tokens,
            cache_read_tokens: month.cache_read_tokens,
        })
        .collect();

    monthly_usage.sort_by(|a, b| a.month.cmp(&b.month));

    let mut model_usage: Vec<ModelUsage> = usage_by_model
        .into_iter()
        .map(|(_, model)| ModelUsage {
            model: model.model,
            input_tokens: model.input_tokens,
            output_tokens: model.output_tokens,
            cached_tokens: model.cached_tokens,
            total_tokens: model.total_tokens,
            cost: format!("{:.4}", model.cost),
            sessions: model.sessions.len(),
            messages: model.messages,
            new_input_tokens: model.new_input_tokens,
            cache_creation_tokens: model.cache_creation_tokens,
            cache_read_tokens: model.cache_read_tokens,
        })
        .collect();

    model_usage.sort_by(|a, b| b.total_tokens.cmp(&a.total_tokens));

    let mut project_data: Vec<ProjectData> = projects
        .into_iter()
        .map(|p| ProjectData {
            name: p.name,
            total_tokens: p.total_tokens,
            total_cost: format!("{:.4}", p.total_cost),
            message_count: p.message_count,
            last_activity: p.last_activity,
        })
        .collect();

    project_data.sort_by(|a, b| {
        b.last_activity
            .as_ref()
            .unwrap_or(&String::new())
            .cmp(a.last_activity.as_ref().unwrap_or(&String::new()))
    });

    Ok(AllProjectData {
        daily_usage,
        monthly_usage,
        model_usage,
        projects: project_data,
    })
}
